### PR TITLE
OAuth 2 support

### PR DESCRIPTION
**Pull request by ngerakines**
This PR introduces OAuth 2.0 support as a client application integrating against a OAuth provider.

### Added
- github.com/writeas/writefreely/db package that contains database and storage utility and helper code
- Database testing utility code in github.com/writeas/writefreely/main_test.go
- OAuth 2.0 HTTP handlers and wired them into /oauth/write.as and /oauth/callback
- The users_oauth and oauth_client_state tables with migrations (v4).

### Changed
- The configuration structure to include OAuth application configuration.

---

- [X] I have signed the [CLA](https://phabricator.write.as/L1)

### CODE REVIEW

File: config/config.go [hunk 1]
```diff
@@ -92,6 +92,15 @@ type (
...
 		UserInvites   string `ini:"user_invites"`
 
+		// OAuth
+		EnableOAuth bool `ini:"enable_oauth"`
+		OAuthProviderAuthLocation    string `ini:"oauth_auth_location"`
+		OAuthProviderTokenLocation   string `ini:"oauth_token_location"`
+		OAuthProviderInspectLocation string `ini:"oauth_inspect_location"`
+		OAuthClientCallbackLocation  string `ini:"oauth_callback_location"`
+		OAuthClientID                string `ini:"oauth_client_id"`
+		OAuthClientSecret            string `ini:"oauth_client_secret"`
```
[the remark this answers is not in the document]
Comment (ngerakines) [reply config/config.go]: I started in that direction with T710. If you're ok with me holding off on those changes for this one, we can review multi-app config in that PR.

[the remark this answers is not in the document]
Comment (thebaer) [reply config/config.go]: Ah, great. Then yeah we can hold off on that.

**Comment by thebaer** [comment]
Thanks. So how can I test and verify this? Do you have sample config values if I'm running Write.as locally?

**Comment by ngerakines** [comment]
> Thanks. So how can I test and verify this? Do you have sample config values if I'm running Write.as locally?

I sent my local setup to you privately.

**Comment by thebaer** [comment]
Perfect -- thanks @ngerakines! Merging now.

**Pull request merged**